{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Complete pipeline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Developed environment \n",
    "\n",
    "- Ubuntu 20.04 LTS (Windows Subsystem for Linux)\n",
    "- Anaconda 3\n",
    "    - Python == 3.8.12 (anaconda)\n",
    "        - biopython == 1.79\n",
    "        - netMHCpan (PyNetMHCpan) == 0.1.5\n",
    "    - fasttree == 2.1.10c\n",
    "    - mafft == 7.487\n",
    "    - mmseqs2 == AVX2\n",
    "- Epitope sequences retrieved from 'iedb.org'\n",
    "- Protein sequences retrieved from 'NCBI influenza database'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import statistics\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "from Bio import AlignIO, Entrez, Phylo\n",
    "from Bio.Align import substitution_matrices\n",
    "\n",
    "from multiprocessing import Pool\n",
    "from itertools import repeat"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# NECESSARY FUNCTIONS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_fa(prot_file):\n",
    "\n",
    "    '''\n",
    "    Read in fasta(.fa) file, separate the id and sequence\n",
    "    Order keeps the same\n",
    "    \n",
    "    [Input]\n",
    "        - prot_file = fasta file to read in\n",
    "    [Return]\n",
    "        - protein_id = list() of protein id (fasta headers)\n",
    "        - protein_seq = list() of protein sequences\n",
    "    '''\n",
    "    seq_id = []\n",
    "    seq = []\n",
    "\n",
    "    with open(prot_file, 'r') as infile:\n",
    "        lines = infile.readlines()\n",
    "        separate_seq = ''\n",
    "        cnt = 0\n",
    "        while cnt != len(lines):\n",
    "            \n",
    "            # Protein ids\n",
    "            if lines[cnt][0] == '>':\n",
    "                seq_id.append(lines[cnt].strip('\\n'))\n",
    "                \n",
    "            # Protein sequences\n",
    "            else:\n",
    "                separate_seq += lines[cnt].strip('\\n')\n",
    "                # make a cut at the end of sequence, or end of the file\n",
    "\n",
    "                if cnt + 1 == len(lines):\n",
    "                    seq.append(separate_seq)\n",
    "                    separate_seq = ''\n",
    "                elif lines[cnt + 1][0] == '>':\n",
    "                    seq.append(separate_seq)\n",
    "                    separate_seq = ''\n",
    "                    \n",
    "            cnt += 1\n",
    "\n",
    "    return seq_id, seq\n",
    "\n",
    "def out_fa(seq_id, seq, output_file):\n",
    "    # *** ASSUME SEQUENCE ID HAVE PREFIX '>' FOR FASTA FILE\n",
    "    with open(output_file, 'w') as outfile:\n",
    "        cnt = 0\n",
    "        while cnt != len(seq):\n",
    "            outfile.write(seq_id[cnt] + '\\n')\n",
    "            outfile.write(seq[cnt] + '\\n')\n",
    "            cnt += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_chrX(seq_id, seq):\n",
    "    filtered_id, filtered_seq = [], []\n",
    "    \n",
    "    for cnt, individ_seq in enumerate(seq):\n",
    "        # count number of ambiguous AAs in the sequence\n",
    "        X_cnt = individ_seq.count(\"X\")\n",
    "\n",
    "        if X_cnt <= 2:\n",
    "            filtered_id.append(seq_id[cnt])\n",
    "            try:\n",
    "                filtered_seq.append(seq[cnt])\n",
    "            except:\n",
    "                print(cnt)\n",
    "                break\n",
    "    \n",
    "    return filtered_id, filtered_seq\n",
    "\n",
    "def collapse_identical_seq(seq_id, seq):\n",
    "    '''\n",
    "    Intake list() of sequences, only transfer first-seen unique sequences.\n",
    "\n",
    "    [Input]\n",
    "        - sequence: list() of sequences\n",
    "    [Return]\n",
    "        - unique_id: list() of unique protein ids\n",
    "        - unique_seq: list() of unique protein seqs\n",
    "    '''\n",
    "    unique_seq = []\n",
    "    unique_id = []\n",
    "    cnt = 0\n",
    "\n",
    "    lengths = [len(s) for s in seq]\n",
    "    mean = statistics.mean(lengths)\n",
    "\n",
    "    while cnt != len(seq_id):\n",
    "        if seq[cnt] not in unique_seq:\n",
    "            unique_id.append(seq_id[cnt])\n",
    "            unique_seq.append(seq[cnt])\n",
    "        cnt +=1\n",
    "    return unique_id, unique_seq\n",
    "\n",
    "def cleanse_short_seq(seq_id, seq):\n",
    "    \n",
    "    lengths = [len(s) for s in seq]\n",
    "    standard = statistics.median(lengths) * 0.9\n",
    "\n",
    "    cleansed_id, cleansed_seq = [], []\n",
    "    cnt = 0\n",
    "\n",
    "    while cnt != len(seq_id):\n",
    "        if len(seq[cnt]) > standard:\n",
    "            cleansed_id.append(seq_id[cnt])\n",
    "            cleansed_seq.append(seq[cnt])\n",
    "        cnt += 1\n",
    "\n",
    "    return cleansed_id, cleansed_seq "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## os.Tools (MSA, cluster, phylotree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_clustering(file, outname, threshold):\n",
    "    '''\n",
    "    Linux command to run MMseqs2 - clustering sequences!\n",
    "    '''\n",
    "    mmseq_run = 'mmseqs easy-cluster {} {} tmp --min-seq-id {} >/dev/null'.format(file, outname, threshold)\n",
    "\n",
    "    # Run mmseqs!\n",
    "    os.system(mmseq_run)\n",
    "\n",
    "def run_alignment(infile, outfile):\n",
    "    '''\n",
    "    Linux command to run MAFFT - multiple sequence alignment!\n",
    "    '''\n",
    "    mafft = 'mafft --auto --thread -1 {} > {}'.format(infile, outfile)\n",
    "\n",
    "    # Run mafft!\n",
    "    os.system(mafft)\n",
    "\n",
    "def run_phylotree(infile, outfile):\n",
    "    '''\n",
    "    Linux command to run FastTree - phylogenetic tree construction!\n",
    "    '''\n",
    "\n",
    "    fasttree = 'fasttree -quote {} > {}'.format(infile, outfile)\n",
    "\n",
    "    # Run fasttree!\n",
    "    os.system(fasttree)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Entrez-fetch subtypes/lineages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_names(seq_id):\n",
    "    from Bio import Entrez\n",
    "\n",
    "    # Always tell NCBI who you are, so that you Avoid any error messages.\n",
    "    Entrez.email = \"JuHyung.Lee@ugent.be\"\n",
    "    handle = Entrez.einfo()\n",
    "    result = handle.read()\n",
    "\n",
    "    accession_id = [id.split()[0][1:] for id in seq_id]\n",
    "    handle = Entrez.efetch(db='protein', id=accession_id, rettype='gb', retmode='text')\n",
    "\n",
    "    # EFETCH-ing multiple accession ids result in continuous strings,\n",
    "    # Where every new id starts by LOCUS information.\n",
    "    # Only the first element is a blank => spliced out.\n",
    "    records = [record.strip() for record in handle.read().split(\"LOCUS\")[1:]]\n",
    "\n",
    "    info_id = []\n",
    "    unique_names = {}\n",
    "\n",
    "    for record in records:\n",
    "        \n",
    "        record = [line.strip() for line in record.split(\"\\n\")]\n",
    "        \n",
    "        # For B virus sequences, strain information only needed\n",
    "        # Some B viruses don't even contain 'strain' section\n",
    "        try:\n",
    "            strain = [i.split(\"=\")[-1].strip(\"\\\"\") for i in record if i.split(\"=\")[0] == '/strain'][0]\n",
    "        except:\n",
    "            strain = [i.split(\"=\")[-1].strip(\"\\\"\") for i in record if i.split(\"=\")[0] == '/organism'][0]\n",
    "\n",
    "        # Remove spaces between city names e.g. Hong Kong -> HongKong\n",
    "        # Phylogenetic tools do not allow spaces in sequence names\n",
    "        strain = strain.replace(\" \", \"\")\n",
    "        \n",
    "        # For A virus sequences, also retrieve subtype info\n",
    "        if strain[0] == 'A':\n",
    "            \n",
    "            # However some accession numbers do not provide explicit 'serotype' section,\n",
    "            # Instead they put all necessary info in 'strain' section.\n",
    "            try:\n",
    "                subtype = [i.split(\"=\")[-1].strip(\"\\\"\") for i in record if i.split(\"=\")[0] == '/serotype'][0]\n",
    "                name = \"{}/{}\".format(strain, subtype)\n",
    "            except:\n",
    "                name = strain\n",
    "\n",
    "        # For B virus sequences. try to retrieve lineages (Yamagata/Victoria), which is not easy...\n",
    "        elif strain[0] == 'B':\n",
    "            try:\n",
    "                search1 = [line.split(\"=\")[0].split(\":\")[-1] for line in record if line.split(\"=\")[0].split(\":\")[0] == 'type']\n",
    "                if search1[0].strip(\" \\\"\") in ['Yamagata', 'Victoria']:\n",
    "                    lineage = search1[0].strip(\" \\\"\")\n",
    "                else:\n",
    "                    search2 = [line.split(\"=\")[-1].split(\":\")[-1] for line in record if 'type' in line.split(\"=\")[-1]]\n",
    "                    if search2[0].strip(\" \\\"\") in ['Yamagata', 'Victoria']:\n",
    "                        lineage = search2[0].strip(\" \\\"\")\n",
    "                name = \"{}/{}\".format(strain, lineage)\n",
    "            except:\n",
    "                name = strain\n",
    "        \n",
    "        else:\n",
    "            name = strain\n",
    "        \n",
    "        # Later on, phylogenetic tree tool (Fasttree) does not accept special characters '()/,\n",
    "        name = name.replace(\"'\", \"_\")\n",
    "        \n",
    "        # Same names could be there\n",
    "        # Give them different numbers\n",
    "        if name not in unique_names:\n",
    "            unique_names[name] = 1\n",
    "        else:\n",
    "            unique_names[name] += 1\n",
    "            name = \"{}_{}\".format(name, unique_names[name])\n",
    "               \n",
    "        info_id.append(\">\" + name)\n",
    "\n",
    "    # Return the final result\n",
    "    return info_id  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Epitopes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### conervancy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [],
   "source": [
    "def match_test(seq1, seq2):\n",
    "    '''\n",
    "    Input:\n",
    "        - Two Amino-Acid sequences of EQUAL lenghth (seq1, seq2)\n",
    "    Output:\n",
    "        - percentage identity between two sequences\n",
    "    '''\n",
    "    cnt = 0\n",
    "    length = len(seq1)\n",
    "    matching_score = 0 # add 1 for every matching parts\n",
    "\n",
    "    ambiguous_residues = {'B':['N', 'D'], 'J':['L', 'I'], 'Z': ['Q', 'E']}\n",
    "    \n",
    "    while cnt != length:\n",
    "        if seq1[cnt] == seq2[cnt]:\n",
    "            matching_score += 1\n",
    "        \n",
    "        elif seq1[cnt] in ambiguous_residues.keys():\n",
    "            if seq2[cnt] in ambiguous_residues[seq1[cnt]]:\n",
    "                matching_score += 0.5\n",
    "            \n",
    "        cnt += 1\n",
    "    \n",
    "    percent_identity = matching_score / length\n",
    "    \n",
    "    return percent_identity\n",
    "    \n",
    "def conservation_analysis(epitope, sequences):\n",
    "    # 1 epitope vs all protein sequences \n",
    "    conservation_results = []\n",
    "\n",
    "    for seq in sequences:\n",
    "            \n",
    "        cnt = 0\n",
    "        max_identity = 0\n",
    "            \n",
    "        while cnt + len(epitope) != len(seq):\n",
    "            \n",
    "            sliding_window = seq[cnt : cnt + len(epitope)]\n",
    "                \n",
    "            # See how the sliding window in protein sequence matches the epitope\n",
    "            # Use function 'def match_test' from above\n",
    "            percent_identity = match_test(sliding_window, epitope)\n",
    "            \n",
    "            # Refresh the best matching score for each protein as the conservancy result\n",
    "            if percent_identity > max_identity:\n",
    "                max_identity = percent_identity\n",
    "            \n",
    "            # If sequences shows 100% at some time, just move onto the next one\n",
    "            # For the sake of computational expense...\n",
    "            if max_identity == 1:\n",
    "                break\n",
    "            cnt += 1\n",
    "                    \n",
    "        conservation_results.append(max_identity)\n",
    "    return conservation_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [],
   "source": [
    "def epitope_conservancy_result_avs(all_results, seq, seqid):\n",
    "    unique_subtypes = set()\n",
    "\n",
    "    for id in seqid:\n",
    "        subtype = re.search(\"H[1-9]*[0-9]*N[1-9]*[0-9]*\", id.split('/')[-1])\n",
    "\n",
    "        if subtype:\n",
    "            unique_subtypes.add(subtype.group(0))\n",
    "\n",
    "    unique_subtypes = list(unique_subtypes)\n",
    "\n",
    "    fraction_80match, fraction_50match, fraction_fullmatch = [], [], []\n",
    "    subtype_match = []\n",
    "\n",
    "    for result in all_results:\n",
    "        # percentage of 100% identity\n",
    "        fraction_fullmatch.append(round((result.count(1.0) / (len(seq))) * 100 , 2))\n",
    "        # Fraction of >80% identity = 1 or 2 AA difference\n",
    "        fraction_80match.append(round(len([i for i in result if i >= 0.8]) / len(seq) * 100, 2))\n",
    "        # Fraction of <50% identity = 1 or 2 AA difference\n",
    "        fraction_50match.append(round(len([i for i in result if i <= 0.5]) / len(seq) * 100, 2))\n",
    "   \n",
    "        # IAV - Number of conserved subtypes\n",
    "        conserved_ids = [seqid[i] for i,j in enumerate(result) if j == 1]\n",
    "        conserved_subtypes = set()\n",
    "        for ids in conserved_ids:\n",
    "            subtype = re.search(\"H[1-9]*[0-9]*N[1-9]*[0-9]*\", ids.split('/')[-1])\n",
    "            if subtype:\n",
    "                conserved_subtypes.add(subtype.group(0))\n",
    "        conserved_subtypes = list(conserved_subtypes)\n",
    "        subtype_match.append(\"{}/{}\".format(len(conserved_subtypes), len(unique_subtypes)))  \n",
    "\n",
    "    data = {'Epitope sequence':epitopes,\n",
    "            'Epitope length': [len(i) for i in epitopes],\n",
    "            'Fraction of 100% matching sequences (%)': fraction_fullmatch,\n",
    "            'Fraction of >80% matching sequences (%)': fraction_80match,\n",
    "            'Fraction of <50% matching sequences (%)': fraction_50match,\n",
    "            '# Subtypes 100% conserved': subtype_match\n",
    "            }\n",
    "\n",
    "    output = pd.DataFrame(data = data)\n",
    "    output.sort_values('Fraction of 100% matching sequences (%)', ascending= False, inplace=True)\n",
    "\n",
    "    return output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [],
   "source": [
    "def epitope_conservancy_result_h(all_results, seq, seqid):\n",
    "\n",
    "    IAV_index = []\n",
    "    IBV_index = []\n",
    "    unique_subtypes = set()\n",
    "\n",
    "    for id in seqid:\n",
    "        if id[1] == 'A':\n",
    "            IAV_index.append(seqid.index(id))\n",
    "            subtype = re.search(\"H[1-9]*[0-9]*N[1-9]*[0-9]*\", id.split('/')[-1])\n",
    "\n",
    "            if subtype:\n",
    "                unique_subtypes.add(subtype.group(0))\n",
    "            \n",
    "        elif id[1] == 'B':\n",
    "            IBV_index.append(seqid.index(id))\n",
    "        elif id[1] == \"I\":\n",
    "            id = id.split(\"Influenza\")\n",
    "\n",
    "            if id[0] == \"A\":\n",
    "                IAV_index.append(seqid.index(id))\n",
    "            elif id[0] == \"B\":\n",
    "                IBV_index.append(seqid.index(id))\n",
    "\n",
    "    unique_subtypes = list(unique_subtypes)\n",
    "    \n",
    "    \n",
    "    fraction_80match, fraction_50match, fraction_fullmatch = [], [], []\n",
    "    fraction_IAVmatch, fraction_IBVmatch = [], []\n",
    "    iav_subtype_match = []\n",
    "\n",
    "    for result in all_results:\n",
    "        # percentage of 100% identity\n",
    "        fraction_fullmatch.append(round((result.count(1.0) / (len(seq))) * 100 , 2))\n",
    "        # Fraction of >80% identity = 1 or 2 AA difference\n",
    "        fraction_80match.append(round(len([i for i in result if i >= 0.8]) / len(seq) * 100, 2))\n",
    "        # Fraction of <50% identity = 1 or 2 AA difference\n",
    "        fraction_50match.append(round(len([i for i in result if i <= 0.5]) / len(seq) * 100, 2))\n",
    "\n",
    "        # Fraction of >80% identity against IAV\n",
    "        fraction_IAVmatch.append(round(len([i for j,i in enumerate(result) if i == 1 and j in IAV_index]) / len(IAV_index) * 100, 2))\n",
    "        \n",
    "        # Number of conserved subtypes\n",
    "        conserved_ids = [seqid[i] for i,j in enumerate(result) if j == 1]\n",
    "        conserved_subtypes = set()\n",
    "        for ids in conserved_ids:\n",
    "            subtype = re.search(\"H[1-9]*[0-9]*N[1-9]*[0-9]*\", ids.split('/')[-1])\n",
    "            if subtype:\n",
    "                conserved_subtypes.add(subtype.group(0))\n",
    "        conserved_subtypes = list(conserved_subtypes)\n",
    "        iav_subtype_match.append(\"{}/{}\".format(len(conserved_subtypes), len(unique_subtypes)))\n",
    "\n",
    "        \n",
    "        # Fraction of >80% identity against IBV\n",
    "        fraction_IBVmatch.append(round(len([i for j,i in enumerate(result) if i == 1 and j in IBV_index]) / len(IBV_index) * 100, 2))\n",
    "\n",
    "    data = {'Epitope sequence':epitopes,\n",
    "            'Epitope length': [len(i) for i in epitopes],\n",
    "            'Fraction of 100% matching sequences (%)': fraction_fullmatch,\n",
    "            'Fraction of >80% matching sequences (%)': fraction_80match,\n",
    "            'Fraction of <50% matching sequences (%)': fraction_50match,\n",
    "            'Fraction of 100% matching in IAV seq (%)': fraction_IAVmatch,\n",
    "            '# Subtypes 100% conserved': iav_subtype_match,\n",
    "            'Fraction of 100% matching in IBV seq (%)': fraction_IBVmatch,\n",
    "            }\n",
    "\n",
    "    output = pd.DataFrame(data = data)\n",
    "    output.sort_values('Fraction of 100% matching sequences (%)', ascending= False, inplace=True)\n",
    "\n",
    "    return output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {},
   "outputs": [],
   "source": [
    "def epitope_conservancy_result_h2(all_results, seq, seqid):\n",
    "\n",
    "    IAV_index = []\n",
    "    IBV_index = []\n",
    "    unique_subtypes = set()\n",
    "\n",
    "    for id in seqid:\n",
    "        if id[1] == 'A':\n",
    "            IAV_index.append(seqid.index(id))\n",
    "            subtype = re.search(\"H[1-9]*[0-9]*N[1-9]*[0-9]*\", id.split('/')[-1])\n",
    "\n",
    "            if subtype:\n",
    "                unique_subtypes.add(subtype.group(0))\n",
    "            \n",
    "        elif id[1] == 'B':\n",
    "            IBV_index.append(seqid.index(id))\n",
    "        elif id[1] == \"I\":\n",
    "            id = id.split(\"Influenza\")\n",
    "\n",
    "            if id[0] == \"A\":\n",
    "                IAV_index.append(seqid.index(id))\n",
    "            elif id[0] == \"B\":\n",
    "                IBV_index.append(seqid.index(id))\n",
    "\n",
    "    unique_subtypes = list(unique_subtypes)\n",
    "    \n",
    "    \n",
    "    fraction_80match, fraction_50match, fraction_fullmatch = [], [], []\n",
    "    fraction_IAVmatch, fraction_IBVmatch = [], []\n",
    "    iav_subtype_match, iav_subtype_partial = [], []\n",
    "\n",
    "    for result in all_results:\n",
    "        # percentage of 100% identity\n",
    "        fraction_fullmatch.append(round((result.count(1.0) / (len(seq))) * 100 , 2))\n",
    "        # Fraction of >80% identity = 1 or 2 AA difference\n",
    "        fraction_80match.append(round(len([i for i in result if i >= 0.8]) / len(seq) * 100, 2))\n",
    "        # Fraction of <50% identity = 1 or 2 AA difference\n",
    "        fraction_50match.append(round(len([i for i in result if i <= 0.5]) / len(seq) * 100, 2))\n",
    "\n",
    "        # Fraction of >80% identity against IAV\n",
    "        fraction_IAVmatch.append(round(len([i for j,i in enumerate(result) if i == 1 and j in IAV_index]) / len(IAV_index) * 100, 2))\n",
    "        \n",
    "        # Number of conserved subtypes\n",
    "        conserved_ids = [seqid[i] for i,j in enumerate(result) if j == 1]\n",
    "        conserved_subtypes = set()\n",
    "        for ids in conserved_ids:\n",
    "            subtype = re.search(\"H[1-9]*[0-9]*N[1-9]*[0-9]*\", ids.split('/')[-1])\n",
    "            if subtype:\n",
    "                conserved_subtypes.add(subtype.group(0))\n",
    "        conserved_subtypes = list(conserved_subtypes)\n",
    "        iav_subtype_match.append(\"{}/{}\".format(len(conserved_subtypes), len(unique_subtypes)))\n",
    "\n",
    "        semi_conserved = [seqid[i] for i,j in enumerate(result) if j >= 0.8]\n",
    "        semi_subtypes = set()\n",
    "        for ids in semi_conserved:\n",
    "            subtype = re.search(\"H[1-9]*[0-9]*N[1-9]*[0-9]*\", ids.split('/')[-1])\n",
    "            if subtype:\n",
    "                semi_subtypes.add(subtype.group(0))\n",
    "\n",
    "        semi_subtypes = list(semi_subtypes)\n",
    "        iav_subtype_partial.append(\"{}/{}\".format(len(semi_subtypes), len(unique_subtypes)))\n",
    "        \n",
    "\n",
    "        \n",
    "        # Fraction of >80% identity against IBV\n",
    "        fraction_IBVmatch.append(round(len([i for j,i in enumerate(result) if i == 1 and j in IBV_index]) / len(IBV_index) * 100, 2))\n",
    "\n",
    "    data = {'Epitope sequence':epitopes,\n",
    "            'Epitope length': [len(i) for i in epitopes],\n",
    "            'Fraction of 100% matching sequences (%)': fraction_fullmatch,\n",
    "            'Fraction of >80% matching sequences (%)': fraction_80match,\n",
    "            'Fraction of <50% matching sequences (%)': fraction_50match,\n",
    "            'Fraction of 100% matching in IAV seq (%)': fraction_IAVmatch,\n",
    "            '# Subtypes 100% conserved': iav_subtype_match,\n",
    "            '# subs 80 conserved': iav_subtype_partial,\n",
    "            'Fraction of 100% matching in IBV seq (%)': fraction_IBVmatch,\n",
    "            }\n",
    "\n",
    "    output = pd.DataFrame(data = data)\n",
    "    output.sort_values('Fraction of 100% matching sequences (%)', ascending= False, inplace=True)\n",
    "\n",
    "    return output"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "def most_frequent(seq):\n",
    "    counter = {}\n",
    "    unique_char = list(set(seq))\n",
    "    seq = list(seq)\n",
    "    \n",
    "    for char in unique_char:\n",
    "        counter[seq.count(char)] = char\n",
    "    \n",
    "    most_freq_value = sorted(counter.keys())[-1]\n",
    "    most_freq_residue = counter[most_freq_value]\n",
    "\n",
    "    return most_freq_residue\n",
    "\n",
    "def consensus_seq(matrix):\n",
    "    '''\n",
    "    Intake a matrix of sequences\n",
    "    Output the consensus sequence.\n",
    "    '''\n",
    "\n",
    "    columns = matrix.shape[1]\n",
    "    consensus = ''\n",
    "\n",
    "    for i in range(columns):\n",
    "        col = columns[:,i]\n",
    "        consensus += most_frequent(col)\n",
    "\n",
    "    return consensus\n",
    "\n",
    "def blosum_scoring(consensus, seq):\n",
    "    # Substitution matrix for scoring sequences\n",
    "    blosum62 = substitution_matrices.read('blosum62.txt')  \n",
    "    score = 0\n",
    "\n",
    "    for AA in seq:\n",
    "        trans = tuple([consensus, AA])\n",
    "        \n",
    "        try:\n",
    "            score += blosum62[trans]\n",
    "        except:\n",
    "            score += blosum62[trans[::-1]]\n",
    "    return score\n",
    "\n",
    "def find_experimental_epitopes(k, alignment, threshold=0.5):\n",
    "    '''\n",
    "    Scan through aligned sequences\n",
    "    to look for new 'possible' epitopes\n",
    "\n",
    "    Input\n",
    "        - k = length of k-mer to look for\n",
    "        - alignment = set of aligned sequences\n",
    "\n",
    "    Returns\n",
    "        - max_score = highest score obtained\n",
    "        - max_index = index of that highest score in the alignment\n",
    "        - max_seq = k-mer sequence with maximum score.\n",
    "    '''\n",
    "    # Record sum of blosum62 score for every 'valid' window (no indels)\n",
    "    score_per_window = []\n",
    "    # Record the maximum sum\n",
    "    std_score_per_window = []\n",
    "\n",
    "    # Alongside, also record the index of starting point of the window.\n",
    "    # For easier tracking of best-scoring regions.\n",
    "    seq_per_window = [] \n",
    "\n",
    "    # Record the positions\n",
    "    positions = []\n",
    "    \n",
    "    cnt = 0\n",
    "        \n",
    "    while cnt + k < alignment.shape[-1]:\n",
    "            \n",
    "        # Extract the window\n",
    "        window = alignment[:, cnt : cnt + k]\n",
    "        window_score = 0\n",
    "        standard_score = 0\n",
    "        consensus_seq = ''\n",
    "        \n",
    "        # We only want amino acid residues\n",
    "        # slid pass the window to next if it contains any indel\n",
    "        if \"-\" not in window:\n",
    "\n",
    "            # For every 'valid' window:\n",
    "            # Iterate through every position (column)\n",
    "            for column_cnt in range(k):\n",
    "                column = window[:, column_cnt]\n",
    "\n",
    "                #Identify the most frequent residue:\n",
    "                consensus = most_frequent(column)\n",
    "                consensus_seq += consensus\n",
    "\n",
    "                #Calculate BLOSUM substitution score per column\n",
    "                window_score += blosum_scoring(consensus, column)\n",
    "\n",
    "                # Exploration: what if I add a conditional?\n",
    "                standard_seq = [consensus] * len(column)\n",
    "                standard_score += blosum_scoring(consensus, standard_seq)\n",
    "\n",
    "        # By default, threshold = 0.8\n",
    "        if window_score > 0 and window_score >= standard_score * threshold:\n",
    "            score_per_window.append(window_score)\n",
    "            std_score_per_window.append(standard_score)\n",
    "            seq_per_window.append(consensus_seq)\n",
    "            \n",
    "            positions.append([cnt+1, cnt+k+1])\n",
    "        cnt += 1\n",
    "\n",
    "    return std_score_per_window, score_per_window, seq_per_window, positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "def exp_epitopes_dataframe(standards, scores, sequences):\n",
    "    \n",
    "    # Process more data out of them\n",
    "    score_fraction = []\n",
    "\n",
    "    cnt = 0\n",
    "\n",
    "    while cnt != len(scores):\n",
    "        frac = round((scores[cnt] / standards[cnt]) * 100, 2)\n",
    "        score_fraction.append(frac)\n",
    "\n",
    "        cnt += 1 \n",
    "    \n",
    "    # Make pandas dataframe\n",
    "    exp_data = {'candidate sequence': sequences, \n",
    "            'length': [len(seq) for seq in sequences],\n",
    "            'substitution score': scores,\n",
    "            'best case': standards,\n",
    "            'score (%)': score_fraction}\n",
    "\n",
    "    exp_dataframe = pd.DataFrame(data = exp_data)\n",
    "    exp_dataframe.sort_values('score (%)', ascending= False, inplace=True)\n",
    "\n",
    "    return exp_dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### non-overlaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def non_overlapping_exp_epitopes(exp_epitopes, epitopes):\n",
    "    '''\n",
    "    A somewhat strict way to look for non-overlapping elements\n",
    "    between known epitopes (retrieved from IEDB) and \n",
    "    detected candidates (conserved regions in MSA).\n",
    "    '''\n",
    "    non_overlap = []\n",
    "    sim_cnt = []\n",
    "\n",
    "    print(len(exp_epitopes) * len(epitopes))\n",
    "    \n",
    "    # Iterate through candidate epitopes (discovered from MSA)\n",
    "    for exp_epitope in exp_epitopes:\n",
    "        similar_cnt = 0\n",
    "\n",
    "        # Iterate through knwon epitopes (IEDB)\n",
    "        for epitope in epitopes:\n",
    "        \n",
    "            # If candidate and known\n",
    "            if epitope in exp_epitope or exp_epitope in epitope:\n",
    "                similar_cnt += 1\n",
    "\n",
    "            # Look for more detailed points\n",
    "            else:\n",
    "                # Unequal lengths\n",
    "                if len(epitope) != len(exp_epitope):\n",
    "                    main = exp_epitope\n",
    "                    slider = epitope\n",
    "                    \n",
    "                    # Swap main/slider if epitope > exp_epitope\n",
    "                    if len(epitope) > len(exp_epitope):\n",
    "                        main = epitope\n",
    "                        slider = exp_epitope\n",
    "                    \n",
    "                    cnt = 0\n",
    "                    k = len(slider)\n",
    "\n",
    "                    while cnt + k < len(main):\n",
    "                        window = main[cnt: cnt+k]\n",
    "                        similarity = match_test(slider, window)\n",
    "\n",
    "                        if similarity > 0.8:\n",
    "                            similar_cnt += 1\n",
    "                        cnt += 1\n",
    "\n",
    "                # Equal lengths\n",
    "                else:\n",
    "                    similarity = match_test(exp_epitope, epitope)\n",
    "                    if similarity > 0.8:\n",
    "                        similar_cnt += 1\n",
    "\n",
    "        sim_cnt.append(similar_cnt)\n",
    "        if similar_cnt == 0:\n",
    "            non_overlap.append(exp_epitope)\n",
    "\n",
    "    if non_overlap:\n",
    "        return similar_cnt, non_overlap\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "def non_overlapping_epitopes_simple(exp_epitopes, known_epitopes):\n",
    "    \"\"\"\n",
    "    Just check whether the candidate epitopes from MSA do not overlap with known epitopes!\n",
    "    \"\"\"\n",
    "\n",
    "    non_overlaps = []\n",
    "\n",
    "    for exp in exp_epitopes:\n",
    "        if exp not in known_epitopes:\n",
    "            non_overlaps.append(exp)\n",
    "    \n",
    "    if non_overlaps:\n",
    "        return non_overlaps\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#  WORKFLOWS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Manual curation of which clustering threshold to use!\n",
    "\n",
    "human_host = {\"HA\": 0.98, \"NA\": 0.98, \"M1\": 0.995, \"M2\": 0.98, \"PA\": 0.99, \n",
    "              \"PB1\":0.99, \"PB2\": 0.99, \"NP\":0.995, \"NS1\":0.98, \"NS2\":0.99}\n",
    "\n",
    "avswine_host = {\"HA\": 0.95, \"NA\": 0.95, \"M1\": 0.99, \"M2\": 0.96, \"PA\": 0.96, \n",
    "                \"PB1\":0.97, \"PB2\": 0.97, \"NP\":0.98, \"NS1\":0.96, \"NS2\":0.97}\n",
    "\n",
    "avswine_tree_host = {\"HA\": 0.87, \"NA\": 0.85, \"M1\": 0.96, \"M2\": 0.91, \"PA\": 0.95, \n",
    "                     \"PB1\": 0.96, \"PB2\": 0.96, \"NP\":0.96, \"NS1\":0.88, \"NS2\":0.92}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 343,
   "metadata": {},
   "outputs": [],
   "source": [
    "prot = \"NS1\"\n",
    "h_clust, avs_clust = human_host[prot], avswine_host[prot]\n",
    "h_clust_str, avs_clust_str = int(h_clust * 100), int(avs_clust * 100)\n",
    "\n",
    "# Conatenate small portion of avian-swine sequences to human sequence\n",
    "# For phylogenetic analysis...\n",
    "tree_avs_clust = avswine_tree_host[prot]\n",
    "tree_avs_clust_str = int(tree_avs_clust * 100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 344,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bringing in the fasta file!\n"
     ]
    }
   ],
   "source": [
    "print(\"Bringing in the fasta file!\")\n",
    "human_protfile = \"../sequences/original_files/{}_AB_allsub.fa\".format(prot, prot)\n",
    "avs_protfile = \"../sequences/original_files/{}_AS_allsub.fa\".format(prot, prot)\n",
    "\n",
    "hseq_id, hseq = read_fa(human_protfile)\n",
    "avseq_id, avseq = read_fa(avs_protfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 345,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Removing sequences with excess X . . .\n",
      " Removing done!\n"
     ]
    }
   ],
   "source": [
    "print(\"Removing sequences with excess X . . .\")\n",
    "hseq_id, hseq = remove_chrX(hseq_id, hseq)\n",
    "avseq_id, avseq = remove_chrX(avseq_id, avseq)\n",
    "print(\" Removing done!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 346,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Remove identical sequences . . .\n",
      "Removing done!\n"
     ]
    }
   ],
   "source": [
    "# Remove identical sequences\n",
    "print(\"Remove identical sequences . . .\")\n",
    "hseq_id, hseq = collapse_identical_seq(hseq_id, hseq)\n",
    "avseq_id, avseq = collapse_identical_seq(avseq_id, avseq)\n",
    "print(\"Removing done!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 347,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Remove excessively short sequences . . .\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "# Remove too-short sequences\n",
    "print(\"Remove excessively short sequences . . .\")\n",
    "hseq_id, hseq = cleanse_short_seq(hseq_id, hseq)\n",
    "avseq_id, avseq = cleanse_short_seq(avseq_id, avseq)\n",
    "print(\"Done!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 348,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Intermediate output - preprocessed file\n",
    "human_collapsed = \"../sequences/processed/{}_AB_clust{}_collapsed.fa\".format(prot, h_clust_str)\n",
    "avswine_collapsed = \"../sequences/processed/{}_AS_clust{}_collapsed.fa\".format(prot, avs_clust_str)\n",
    "out_fa(hseq_id, hseq, human_collapsed)\n",
    "out_fa(avseq_id, avseq, avswine_collapsed)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 349,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Perform clustering\n",
    "human_cluster_output_file = \"../sequences/clustered/{}_AB_cluster{}\".format(prot, h_clust_str)\n",
    "avswine_cluster_output_file = \"../sequences/clustered/{}_AS_cluster{}\".format(prot, avs_clust_str)\n",
    "avs_tree_cluster_output_file = \"../sequences/clustered/{}_AS_cluster{}_extra\".format(prot, tree_avs_clust_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 350,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Clusterng human-host sequences . . .\n",
      "Clusterng avian-and-swine-host sequences . . .\n",
      "Clusterng extra small portion of avian-and-swine-host sequences . . .\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 350,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"Clusterng human-host sequences . . .\")\n",
    "run_clustering(file=human_collapsed, outname=human_cluster_output_file, threshold=h_clust)\n",
    "\n",
    "print(\"Clusterng avian-and-swine-host sequences . . .\")\n",
    "run_clustering(file=avswine_collapsed, outname=avswine_cluster_output_file, threshold=avs_clust)\n",
    "\n",
    "print(\"Clusterng extra small portion of avian-and-swine-host sequences . . .\")\n",
    "run_clustering(file=avswine_collapsed, outname=avs_tree_cluster_output_file, threshold=tree_avs_clust)\n",
    "\n",
    "# Other files from mmseqs output not needed... remove them\n",
    "os.system(\"rm ../sequences/clustered/{}_A*_cluster*_all_seqs*.fasta\".format(prot))\n",
    "os.system(\"rm ../sequences/clustered/{}_A*_cluster*_cluster*.tsv\".format(prot))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 351,
   "metadata": {},
   "outputs": [],
   "source": [
    "human_cluster_repseq_file = \"{}_rep_seq.fasta\".format(human_cluster_output_file)\n",
    "avswine_cluster_repseq_file = \"{}_rep_seq.fasta\".format(avswine_cluster_output_file)\n",
    "avswine_cluster_extra_repseq = \"{}_rep_seq.fasta\".format(avs_tree_cluster_output_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 352,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read clustered files\n",
    "hseq_id, hseq = read_fa(human_cluster_repseq_file)\n",
    "avseq_id, avseq = read_fa(avswine_cluster_repseq_file)\n",
    "extra_id, extra_seq = read_fa(avswine_cluster_extra_repseq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fetch names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 353,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching human-host sequences information from Entrez . . .\n",
      "Fetching avian and swine sequences information from Entrez . . .\n",
      "Fetching extra portion of avian and swine sequences information from Entrez . . .\n",
      "DONE!\n"
     ]
    }
   ],
   "source": [
    "# RE-Formalize the fasta seq names\n",
    "print(\"Fetching human-host sequences information from Entrez . . .\")\n",
    "hseq_id = fetch_names(hseq_id)\n",
    "\n",
    "print(\"Fetching avian and swine sequences information from Entrez . . .\")\n",
    "avseq_id = fetch_names(avseq_id)    \n",
    "\n",
    "print(\"Fetching extra portion of avian and swine sequences information from Entrez . . .\")\n",
    "extra_id = fetch_names(extra_id)\n",
    "\n",
    "print(\"DONE!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 354,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Intermediate output - name-processed file\n",
    "human_processed = \"../sequences/processed/{}_AB_clust{}_processed.fa\".format(prot, h_clust_str)\n",
    "avswine_processed = \"../sequences/processed/{}_AS_clust{}_processed.fa\".format(prot, avs_clust_str)\n",
    "extra_processed = \"../sequences/processed/{}_AS_clust{}_processed_extra.fa\".format(prot, tree_avs_clust_str)\n",
    "out_fa(hseq_id, hseq, human_processed)\n",
    "out_fa(avseq_id, avseq, avswine_processed)\n",
    "out_fa(extra_id, extra_seq, extra_processed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 355,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 355,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concat_seq = \"../sequences/processed/{}_concat_procssed.fasta\".format(prot)\n",
    "os.system(\"cat {} {} > {}\".format(human_processed,\n",
    "                                  extra_processed,\n",
    "                                  concat_seq))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MSA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 356,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Performing mafft MSA . . .\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "OS = linux\n",
      "The number of physical cores =  6\n",
      "nthread = 6\n",
      "nthreadpair = 6\n",
      "nthreadtb = 6\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "\n",
      "There are 20 ambiguous characters.\n",
      "  301 / 315 (thread    1)\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=0) ... \n",
      "  310 / 315\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/2... \n",
      "STEP   314 / 314 (thread    1) h\n",
      "done.\n",
      "\n",
      "Making a distance matrix from msa.. \n",
      "  300 / 315 (thread    5)\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "  310 / 315\n",
      "done.\n",
      "\n",
      "Progressive alignment 2/2... \n",
      "STEP   314 / 314 (thread    1) h\n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.487\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "6 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "  300 / 315 (thread    4)dndpre (aa) Version 7.487\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "6 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 6\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "rescale = 1\n",
      "\n",
      "  310 / 315\n",
      "Segment   1/  7    1-  51\n",
      "\n",
      "Converged.\n",
      "Segment   2/  7   51- 169\n",
      "\n",
      "Converged.\n",
      "Segment   3/  7  169- 203\n",
      "\n",
      "Converged.\n",
      "\n",
      "Reached 2\n",
      "Segment   4/  7  205- 210\n",
      "\n",
      "Converged.\n",
      "Segment   5/  7  210- 251\n",
      "\n",
      "Converged.\n",
      "Segment   6/  7  251- 316\n",
      "\n",
      "Converged.\n",
      "\n",
      "Reached 2\n",
      "Segment   7/  7  316- 379\n",
      "\n",
      "Converged.\n",
      "done\n",
      "dvtditr (aa) Version 7.487\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "6 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-NS-i (Standard)\n",
      " Iterative refinement method (max. 2 iterations)\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Performing avswine mafft\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "OS = linux\n",
      "The number of physical cores =  6\n",
      "nthread = 6\n",
      "nthreadpair = 6\n",
      "nthreadtb = 6\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "\n",
      "There are 22 ambiguous characters.\n",
      "  601 / 646 (thread    5)\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=0) ... \n",
      "  640 / 646\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/2... \n",
      "STEP   501 / 645 (thread    3)\n",
      "Reallocating..done. *alloclen = 1475\n",
      "STEP   601 / 645 (thread    1)\n",
      "done.\n",
      "\n",
      "Making a distance matrix from msa.. \n",
      "  600 / 646 (thread    3)\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "  640 / 646\n",
      "done.\n",
      "\n",
      "Progressive alignment 2/2... \n",
      "STEP   500 / 645 (thread    5)\n",
      "Reallocating..done. *alloclen = 1475\n",
      "STEP   601 / 645 (thread    2)\n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.487\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "6 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-NS-2 (Fast but rough)\n",
      " Progressive method (guide trees were built 2 times.)\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Performing MSA on concatenated . . .\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "OS = linux\n",
      "The number of physical cores =  6\n",
      "nthread = 6\n",
      "nthreadpair = 6\n",
      "nthreadtb = 6\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "\n",
      "There are 20 ambiguous characters.\n",
      "  301 / 347 (thread    5)\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=0) ... \n",
      "  340 / 347\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/2... \n",
      "STEP   346 / 346 (thread    3) h\n",
      "done.\n",
      "\n",
      "Making a distance matrix from msa.. \n",
      "  300 / 347 (thread    2)\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "  340 / 347\n",
      "done.\n",
      "\n",
      "Progressive alignment 2/2... \n",
      "STEP   346 / 346 (thread    0) h\n",
      "done.\n",
      "\n",
      "disttbfast (aa) Version 7.487\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "6 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "  300 / 347 (thread    4)dndpre (aa) Version 7.487\n",
      "alg=X, model=BLOSUM62, 1.53, +0.12, -0.00, noshift, amax=0.0\n",
      "6 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 6\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "rescale = 1\n",
      "\n",
      "  340 / 347\n",
      "Segment   1/  7    1-  51\n",
      "\n",
      "Converged.\n",
      "\n",
      "Reached 2\n",
      "Segment   2/  7   51- 169\n",
      "\n",
      "Converged.\n",
      "Segment   3/  7  169- 204\n",
      "\n",
      "Converged.\n",
      "\n",
      "Reached 2\n",
      "Segment   4/  7  206- 211\n",
      "\n",
      "Converged.\n",
      "Segment   5/  7  211- 252\n",
      "\n",
      "Converged.\n",
      "Segment   6/  7  252- 317\n",
      "\n",
      "Converged.\n",
      "\n",
      "Reached 2\n",
      "Segment   7/  7  317- 380\n",
      "\n",
      "Converged.\n",
      "\n",
      "Reached 2\n",
      "done\n",
      "dvtditr (aa) Version 7.487\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "6 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-NS-i (Standard)\n",
      " Iterative refinement method (max. 2 iterations)\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MSA done!\n"
     ]
    }
   ],
   "source": [
    "# Perform Multiple Sequence Alignment\n",
    "human_msa = \"../sequences/msa/msa_{}_AB_clust{}.fa\".format(prot, h_clust_str)\n",
    "avswine_msa =  \"../sequences/msa/msa_{}_AS_clust{}.fa\".format(prot, avs_clust_str)\n",
    "concat_msa = \"../sequences/msa/msa_{}_concat.fa\".format(prot)\n",
    "\n",
    "print(\"Performing mafft MSA . . .\")\n",
    "run_alignment(infile=human_processed, outfile=human_msa)\n",
    "\n",
    "print(\"Performing avswine mafft\")\n",
    "run_alignment(infile=avswine_processed, outfile=avswine_msa)\n",
    "\n",
    "print(\"Performing MSA on concatenated . . .\")\n",
    "run_alignment(infile=concat_seq, outfile=concat_msa)\n",
    "\n",
    "print(\"MSA done!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 357,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Phylogenetic tree construction . . .\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "FastTree Version 2.1.10 Double precision (No SSE3)\n",
      "Alignment: ../sequences/msa/msa_NS1_concat.fa\n",
      "Amino acid distances: BLOSUM45 Joins: balanced Support: SH-like 1000\n",
      "Search: Normal +NNI +SPR (2 rounds range 10) +ML-NNI opt-each=1\n",
      "TopHits: 1.00*sqrtN close=default refresh=0.80\n",
      "ML Model: Jones-Taylor-Thorton, CAT approximation with 20 rate categories\n",
      "Ignored unknown character J (seen 1 times)\n",
      "Ignored unknown character X (seen 20 times)\n",
      "      0.10 seconds: Joined    300 of    344\n",
      "Initial topology in 0.11 seconds\n",
      "Refining topology: 34 rounds ME-NNIs, 2 rounds ME-SPRs, 17 rounds ML-NNIs\n",
      "      0.21 seconds: SPR round   1 of   2, 101 of 692 nodes\n",
      "      0.33 seconds: SPR round   1 of   2, 401 of 692 nodes\n",
      "      0.47 seconds: ME NNI round 12 of 34, 1 of 345 splits\n",
      "      0.58 seconds: SPR round   2 of   2, 201 of 692 nodes\n",
      "      0.71 seconds: SPR round   2 of   2, 501 of 692 nodes\n",
      "Total branch-length 9.107 after 0.83 sec\n",
      "      0.83 seconds: ML Lengths 1 of 345 splits\n",
      "      0.97 seconds: ML Lengths 101 of 345 splits\n",
      "      1.11 seconds: ML Lengths 201 of 345 splits\n",
      "      1.25 seconds: ML Lengths 301 of 345 splits\n",
      "      1.80 seconds: ML NNI round 1 of 17, 101 of 345 splits, 58 changes (max delta 3.353)\n",
      "      2.30 seconds: ML NNI round 1 of 17, 201 of 345 splits, 87 changes (max delta 5.917)\n",
      "      2.81 seconds: ML NNI round 1 of 17, 301 of 345 splits, 129 changes (max delta 14.246)\n",
      "ML-NNI round 1: LogLk = -15282.163 NNIs 143 max delta 14.25 Time 3.01\n",
      "      3.04 seconds: Site likelihoods with rate category 1 of 20\n",
      "      3.16 seconds: Site likelihoods with rate category 5 of 20\n",
      "      3.29 seconds: Site likelihoods with rate category 9 of 20\n",
      "      3.41 seconds: Site likelihoods with rate category 13 of 20\n",
      "      3.53 seconds: Site likelihoods with rate category 17 of 20\n",
      "Switched to using 20 rate categories (CAT approximation)\n",
      "Rate categories were divided by 0.878 so that average rate = 1.0\n",
      "CAT-based log-likelihoods may not be comparable across runs\n",
      "Use -gamma for approximate but comparable Gamma(20) log-likelihoods\n",
      "      3.68 seconds: ML NNI round 2 of 17, 1 of 345 splits\n",
      "      4.12 seconds: ML NNI round 2 of 17, 101 of 345 splits, 25 changes (max delta 4.533)\n",
      "      4.59 seconds: ML NNI round 2 of 17, 201 of 345 splits, 62 changes (max delta 4.533)\n",
      "      5.06 seconds: ML NNI round 2 of 17, 301 of 345 splits, 101 changes (max delta 6.553)\n",
      "ML-NNI round 2: LogLk = -14450.399 NNIs 124 max delta 6.55 Time 5.29\n",
      "      5.29 seconds: ML NNI round 3 of 17, 1 of 345 splits\n",
      "      5.75 seconds: ML NNI round 3 of 17, 101 of 345 splits, 35 changes (max delta 4.114)\n",
      "      6.20 seconds: ML NNI round 3 of 17, 201 of 345 splits, 67 changes (max delta 4.114)\n",
      "      6.66 seconds: ML NNI round 3 of 17, 301 of 345 splits, 110 changes (max delta 4.114)\n",
      "ML-NNI round 3: LogLk = -14426.282 NNIs 111 max delta 4.11 Time 6.67\n",
      "      7.19 seconds: ML NNI round 4 of 17, 101 of 345 splits, 53 changes (max delta 0.000)\n",
      "      7.63 seconds: ML NNI round 4 of 17, 201 of 345 splits, 77 changes (max delta 3.783)\n",
      "ML-NNI round 4: LogLk = -14418.684 NNIs 109 max delta 3.78 Time 8.01\n",
      "      8.01 seconds: ML NNI round 5 of 17, 1 of 345 splits\n",
      "      8.46 seconds: ML NNI round 5 of 17, 101 of 345 splits, 29 changes (max delta 3.242)\n",
      "      8.90 seconds: ML NNI round 5 of 17, 201 of 345 splits, 65 changes (max delta 3.242)\n",
      "ML-NNI round 5: LogLk = -14413.272 NNIs 101 max delta 3.24 Time 9.30\n",
      "      9.30 seconds: ML NNI round 6 of 17, 1 of 345 splits\n",
      "      9.83 seconds: ML NNI round 6 of 17, 101 of 345 splits, 55 changes (max delta 3.690)\n",
      "     10.23 seconds: ML NNI round 6 of 17, 201 of 345 splits, 79 changes (max delta 3.690)\n",
      "ML-NNI round 6: LogLk = -14405.909 NNIs 82 max delta 3.69 Time 10.33\n",
      "     10.75 seconds: ML NNI round 7 of 17, 101 of 345 splits, 34 changes (max delta 0.000)\n",
      "ML-NNI round 7: LogLk = -14405.888 NNIs 66 max delta 0.00 Time 11.10\n",
      "Turning off heuristics for final round of ML NNIs (converged)\n",
      "     11.09 seconds: ML NNI round 8 of 17, 1 of 345 splits\n",
      "     11.63 seconds: ML NNI round 8 of 17, 101 of 345 splits, 53 changes (max delta 0.000)\n",
      "     12.12 seconds: ML NNI round 8 of 17, 201 of 345 splits, 73 changes (max delta 0.000)\n",
      "     12.62 seconds: ML NNI round 8 of 17, 301 of 345 splits, 96 changes (max delta 0.722)\n",
      "ML-NNI round 8: LogLk = -14403.415 NNIs 102 max delta 0.72 Time 12.82 (final)\n",
      "     12.81 seconds: ML Lengths 1 of 345 splits\n",
      "     12.95 seconds: ML Lengths 101 of 345 splits\n",
      "     13.07 seconds: ML Lengths 201 of 345 splits\n",
      "     13.20 seconds: ML Lengths 301 of 345 splits\n",
      "Optimize all lengths: LogLk = -14402.834 Time 13.27\n",
      "     13.73 seconds: ML split tests for    100 of    344 internal splits\n",
      "     14.18 seconds: ML split tests for    200 of    344 internal splits\n",
      "     14.60 seconds: ML split tests for    300 of    344 internal splits\n",
      "Total time: 14.78 seconds Unique: 347/347 Bad splits: 3/344 Worst delta-LogLk 3.587\n"
     ]
    }
   ],
   "source": [
    "# Perform Phylogenetic Tree construction\n",
    "print(\"Phylogenetic tree construction . . .\")\n",
    "concat_phylo = \"../sequences/trees/tree_{}_h{}_as{}.new\".format(prot, h_clust_str, tree_avs_clust_str)\n",
    "run_phylotree(infile=concat_msa, outfile=concat_phylo)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Epitopes conservancy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 364,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Re-load protein sequences\n",
    "prot = \"NS1\"\n",
    "h_clust, avs_clust = human_host[prot], avswine_host[prot]\n",
    "h_clust_str, avs_clust_str = int(h_clust * 100), int(avs_clust * 100)\n",
    "\n",
    "human_processed = \"../sequences/processed/{}_AB_clust{}_processed.fa\".format(prot, h_clust_str)\n",
    "avswine_processed = \"../sequences/processed/{}_AS_clust{}_processed.fa\".format(prot, avs_clust_str)\n",
    "\n",
    "hseq_id, hseq = read_fa(human_processed)\n",
    "avseq_id, avseq = read_fa(avswine_processed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 365,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now we look into known epitopes.. how they conserved?\n",
    "epitope_file = \"../sequences/epitopes/{}_AB_epitopes.csv\".format(prot)\n",
    "epitopes_csv = pd.read_csv(epitope_file, header=1)\n",
    "epitopes = epitopes_csv[\"Description\"].tolist()\n",
    "\n",
    "# Some epitopes are in form of: SYQKEIQAKETMK + DEAM(Q3)\n",
    "# The addition of unkown words should be gone.\n",
    "epitopes = [i.split(\"+\")[0].strip() for i in epitopes]\n",
    "\n",
    "# I will only look for epitopes of length 9~11, typical for presentation of MHC-I alleles\n",
    "epitopes = [i for i in epitopes if 9 <= len(i) <= 11]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 366,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing conservancy in human host sequences . . .\n",
      "Processing conservancy in avian/swine host sequences . . .\n"
     ]
    }
   ],
   "source": [
    "from multiprocessing import Pool\n",
    "from itertools import repeat\n",
    "\n",
    "print(\"Processing conservancy in human host sequences . . .\")\n",
    "with Pool(processes=6) as pool:\n",
    "    h_conservancy = pool.starmap(conservation_analysis, zip(epitopes, repeat(hseq)))\n",
    "\n",
    "print(\"Processing conservancy in avian/swine host sequences . . .\")\n",
    "with Pool(processes=6) as pool:\n",
    "    avs_conservancy = pool.starmap(conservation_analysis, zip(epitopes, repeat(avseq)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 367,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Summarizing human-sequences epitope conservancy results . . .\n",
      "Summarizing avian-swine sequences epitope conservancy results . . .\n"
     ]
    }
   ],
   "source": [
    "print(\"Summarizing human-sequences epitope conservancy results . . .\")\n",
    "h_conservancy_table = epitope_conservancy_result_h2(h_conservancy, hseq, hseq_id)\n",
    "print(\"Summarizing avian-swine sequences epitope conservancy results . . .\")\n",
    "avs_conservancy_table = epitope_conservancy_result_avs(avs_conservancy, avseq, avseq_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 368,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Epitope sequence</th>\n",
       "      <th>Epitope length</th>\n",
       "      <th>Fraction of 100% matching sequences (%)</th>\n",
       "      <th>Fraction of &gt;80% matching sequences (%)</th>\n",
       "      <th>Fraction of &lt;50% matching sequences (%)</th>\n",
       "      <th>Fraction of 100% matching in IAV seq (%)</th>\n",
       "      <th># Subtypes 100% conserved</th>\n",
       "      <th># subs 80 conserved</th>\n",
       "      <th>Fraction of 100% matching in IBV seq (%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>GVLIGGLEW</td>\n",
       "      <td>9</td>\n",
       "      <td>71.75</td>\n",
       "      <td>81.59</td>\n",
       "      <td>15.24</td>\n",
       "      <td>84.62</td>\n",
       "      <td>8/10</td>\n",
       "      <td>10/10</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>DRLRRDQKS</td>\n",
       "      <td>9</td>\n",
       "      <td>53.02</td>\n",
       "      <td>83.17</td>\n",
       "      <td>15.24</td>\n",
       "      <td>61.54</td>\n",
       "      <td>9/10</td>\n",
       "      <td>10/10</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GEISPLPSL</td>\n",
       "      <td>9</td>\n",
       "      <td>44.13</td>\n",
       "      <td>83.17</td>\n",
       "      <td>15.24</td>\n",
       "      <td>50.77</td>\n",
       "      <td>9/10</td>\n",
       "      <td>9/10</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>MSRDWFMLM</td>\n",
       "      <td>9</td>\n",
       "      <td>39.05</td>\n",
       "      <td>50.16</td>\n",
       "      <td>15.24</td>\n",
       "      <td>45.77</td>\n",
       "      <td>8/10</td>\n",
       "      <td>8/10</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>IILKANFSV</td>\n",
       "      <td>9</td>\n",
       "      <td>29.84</td>\n",
       "      <td>64.13</td>\n",
       "      <td>15.24</td>\n",
       "      <td>35.00</td>\n",
       "      <td>8/10</td>\n",
       "      <td>10/10</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>RLRRDQRSL</td>\n",
       "      <td>9</td>\n",
       "      <td>28.25</td>\n",
       "      <td>82.86</td>\n",
       "      <td>0.63</td>\n",
       "      <td>34.23</td>\n",
       "      <td>4/10</td>\n",
       "      <td>9/10</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ETIVLLRAF</td>\n",
       "      <td>9</td>\n",
       "      <td>22.86</td>\n",
       "      <td>33.02</td>\n",
       "      <td>15.24</td>\n",
       "      <td>27.69</td>\n",
       "      <td>2/10</td>\n",
       "      <td>4/10</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AIMEKNIML</td>\n",
       "      <td>9</td>\n",
       "      <td>20.63</td>\n",
       "      <td>37.46</td>\n",
       "      <td>16.19</td>\n",
       "      <td>25.00</td>\n",
       "      <td>3/10</td>\n",
       "      <td>4/10</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>IMLKANFSV</td>\n",
       "      <td>9</td>\n",
       "      <td>17.78</td>\n",
       "      <td>70.79</td>\n",
       "      <td>15.24</td>\n",
       "      <td>20.38</td>\n",
       "      <td>4/10</td>\n",
       "      <td>10/10</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>QELSDAPFL</td>\n",
       "      <td>9</td>\n",
       "      <td>16.19</td>\n",
       "      <td>53.33</td>\n",
       "      <td>15.87</td>\n",
       "      <td>19.62</td>\n",
       "      <td>2/10</td>\n",
       "      <td>9/10</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>IASVPTSRY</td>\n",
       "      <td>9</td>\n",
       "      <td>15.24</td>\n",
       "      <td>20.00</td>\n",
       "      <td>22.54</td>\n",
       "      <td>17.69</td>\n",
       "      <td>2/10</td>\n",
       "      <td>3/10</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ALDYPGQDRL</td>\n",
       "      <td>10</td>\n",
       "      <td>14.92</td>\n",
       "      <td>15.24</td>\n",
       "      <td>84.76</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0/10</td>\n",
       "      <td>0/10</td>\n",
       "      <td>97.37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>NVLSLRVLV</td>\n",
       "      <td>9</td>\n",
       "      <td>14.29</td>\n",
       "      <td>15.24</td>\n",
       "      <td>84.76</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0/10</td>\n",
       "      <td>0/10</td>\n",
       "      <td>94.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>SQFGQEHRL</td>\n",
       "      <td>9</td>\n",
       "      <td>14.29</td>\n",
       "      <td>14.92</td>\n",
       "      <td>85.08</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0/10</td>\n",
       "      <td>0/10</td>\n",
       "      <td>97.37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>SQFGQEHRL</td>\n",
       "      <td>9</td>\n",
       "      <td>14.29</td>\n",
       "      <td>14.92</td>\n",
       "      <td>85.08</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0/10</td>\n",
       "      <td>0/10</td>\n",
       "      <td>97.37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>VLSQFGQEHRL</td>\n",
       "      <td>11</td>\n",
       "      <td>14.29</td>\n",
       "      <td>14.92</td>\n",
       "      <td>85.08</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0/10</td>\n",
       "      <td>0/10</td>\n",
       "      <td>97.37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>KRKMARTAR</td>\n",
       "      <td>9</td>\n",
       "      <td>13.02</td>\n",
       "      <td>30.16</td>\n",
       "      <td>40.32</td>\n",
       "      <td>15.38</td>\n",
       "      <td>2/10</td>\n",
       "      <td>4/10</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AIMDKNIIL</td>\n",
       "      <td>9</td>\n",
       "      <td>12.70</td>\n",
       "      <td>33.65</td>\n",
       "      <td>15.24</td>\n",
       "      <td>15.00</td>\n",
       "      <td>5/10</td>\n",
       "      <td>7/10</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>IMDKNIILKA</td>\n",
       "      <td>10</td>\n",
       "      <td>12.70</td>\n",
       "      <td>63.17</td>\n",
       "      <td>15.24</td>\n",
       "      <td>15.00</td>\n",
       "      <td>5/10</td>\n",
       "      <td>9/10</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>AYDQSGRLV</td>\n",
       "      <td>9</td>\n",
       "      <td>11.75</td>\n",
       "      <td>14.29</td>\n",
       "      <td>84.13</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0/10</td>\n",
       "      <td>0/10</td>\n",
       "      <td>81.58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>MTIASVPTSRY</td>\n",
       "      <td>11</td>\n",
       "      <td>11.75</td>\n",
       "      <td>24.44</td>\n",
       "      <td>24.76</td>\n",
       "      <td>13.46</td>\n",
       "      <td>2/10</td>\n",
       "      <td>6/10</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MASTPASRY</td>\n",
       "      <td>9</td>\n",
       "      <td>11.75</td>\n",
       "      <td>37.14</td>\n",
       "      <td>21.59</td>\n",
       "      <td>14.23</td>\n",
       "      <td>2/10</td>\n",
       "      <td>4/10</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>MTIASVPTSR</td>\n",
       "      <td>10</td>\n",
       "      <td>11.75</td>\n",
       "      <td>24.44</td>\n",
       "      <td>26.67</td>\n",
       "      <td>13.46</td>\n",
       "      <td>2/10</td>\n",
       "      <td>6/10</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>MPRQKIIGPL</td>\n",
       "      <td>10</td>\n",
       "      <td>11.43</td>\n",
       "      <td>17.46</td>\n",
       "      <td>17.46</td>\n",
       "      <td>13.46</td>\n",
       "      <td>3/10</td>\n",
       "      <td>3/10</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>ILNSLFERL</td>\n",
       "      <td>9</td>\n",
       "      <td>10.48</td>\n",
       "      <td>15.24</td>\n",
       "      <td>84.44</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0/10</td>\n",
       "      <td>0/10</td>\n",
       "      <td>73.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AIMEKNIIL</td>\n",
       "      <td>9</td>\n",
       "      <td>5.40</td>\n",
       "      <td>44.76</td>\n",
       "      <td>15.56</td>\n",
       "      <td>6.54</td>\n",
       "      <td>2/10</td>\n",
       "      <td>7/10</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>AVMEKNIVL</td>\n",
       "      <td>9</td>\n",
       "      <td>2.86</td>\n",
       "      <td>11.75</td>\n",
       "      <td>17.46</td>\n",
       "      <td>3.08</td>\n",
       "      <td>2/10</td>\n",
       "      <td>2/10</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>VPASRYLTDM</td>\n",
       "      <td>10</td>\n",
       "      <td>2.54</td>\n",
       "      <td>63.17</td>\n",
       "      <td>15.24</td>\n",
       "      <td>3.08</td>\n",
       "      <td>3/10</td>\n",
       "      <td>8/10</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>FLWHVRKRV</td>\n",
       "      <td>9</td>\n",
       "      <td>2.54</td>\n",
       "      <td>34.60</td>\n",
       "      <td>15.24</td>\n",
       "      <td>3.08</td>\n",
       "      <td>1/10</td>\n",
       "      <td>8/10</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>FMDPSAGIEGF</td>\n",
       "      <td>11</td>\n",
       "      <td>2.54</td>\n",
       "      <td>14.92</td>\n",
       "      <td>84.76</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0/10</td>\n",
       "      <td>0/10</td>\n",
       "      <td>15.79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AIMDKNIML</td>\n",
       "      <td>9</td>\n",
       "      <td>1.90</td>\n",
       "      <td>38.41</td>\n",
       "      <td>15.24</td>\n",
       "      <td>1.54</td>\n",
       "      <td>2/10</td>\n",
       "      <td>8/10</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ALKMTMASV</td>\n",
       "      <td>9</td>\n",
       "      <td>1.27</td>\n",
       "      <td>20.95</td>\n",
       "      <td>25.40</td>\n",
       "      <td>1.54</td>\n",
       "      <td>1/10</td>\n",
       "      <td>5/10</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>AIMEKNIVLK</td>\n",
       "      <td>10</td>\n",
       "      <td>0.95</td>\n",
       "      <td>59.37</td>\n",
       "      <td>16.19</td>\n",
       "      <td>1.15</td>\n",
       "      <td>1/10</td>\n",
       "      <td>8/10</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>AIMEKNIVL</td>\n",
       "      <td>9</td>\n",
       "      <td>0.95</td>\n",
       "      <td>33.02</td>\n",
       "      <td>16.19</td>\n",
       "      <td>1.15</td>\n",
       "      <td>1/10</td>\n",
       "      <td>3/10</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>AIVDKNITL</td>\n",
       "      <td>9</td>\n",
       "      <td>0.63</td>\n",
       "      <td>2.22</td>\n",
       "      <td>17.46</td>\n",
       "      <td>0.77</td>\n",
       "      <td>1/10</td>\n",
       "      <td>3/10</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>IPKQKVAGPL</td>\n",
       "      <td>10</td>\n",
       "      <td>0.32</td>\n",
       "      <td>56.83</td>\n",
       "      <td>18.10</td>\n",
       "      <td>0.38</td>\n",
       "      <td>1/10</td>\n",
       "      <td>8/10</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>LPSLPGHTA</td>\n",
       "      <td>9</td>\n",
       "      <td>0.00</td>\n",
       "      <td>44.44</td>\n",
       "      <td>15.56</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0/10</td>\n",
       "      <td>9/10</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Epitope sequence  Epitope length  Fraction of 100% matching sequences (%)  \\\n",
       "15        GVLIGGLEW               9                                    71.75   \n",
       "5         DRLRRDQKS               9                                    53.02   \n",
       "7         GEISPLPSL               9                                    44.13   \n",
       "27        MSRDWFMLM               9                                    39.05   \n",
       "22        IILKANFSV               9                                    29.84   \n",
       "13        RLRRDQRSL               9                                    28.25   \n",
       "14        ETIVLLRAF               9                                    22.86   \n",
       "3         AIMEKNIML               9                                    20.63   \n",
       "16        IMLKANFSV               9                                    17.78   \n",
       "21        QELSDAPFL               9                                    16.19   \n",
       "26        IASVPTSRY               9                                    15.24   \n",
       "29       ALDYPGQDRL              10                                    14.92   \n",
       "33        NVLSLRVLV               9                                    14.29   \n",
       "34        SQFGQEHRL               9                                    14.29   \n",
       "35        SQFGQEHRL               9                                    14.29   \n",
       "36      VLSQFGQEHRL              11                                    14.29   \n",
       "17        KRKMARTAR               9                                    13.02   \n",
       "0         AIMDKNIIL               9                                    12.70   \n",
       "8        IMDKNIILKA              10                                    12.70   \n",
       "30        AYDQSGRLV               9                                    11.75   \n",
       "20      MTIASVPTSRY              11                                    11.75   \n",
       "11        MASTPASRY               9                                    11.75   \n",
       "19       MTIASVPTSR              10                                    11.75   \n",
       "18       MPRQKIIGPL              10                                    11.43   \n",
       "32        ILNSLFERL               9                                    10.48   \n",
       "2         AIMEKNIIL               9                                     5.40   \n",
       "24        AVMEKNIVL               9                                     2.86   \n",
       "12       VPASRYLTDM              10                                     2.54   \n",
       "6         FLWHVRKRV               9                                     2.54   \n",
       "31      FMDPSAGIEGF              11                                     2.54   \n",
       "1         AIMDKNIML               9                                     1.90   \n",
       "4         ALKMTMASV               9                                     1.27   \n",
       "28       AIMEKNIVLK              10                                     0.95   \n",
       "23        AIMEKNIVL               9                                     0.95   \n",
       "25        AIVDKNITL               9                                     0.63   \n",
       "9        IPKQKVAGPL              10                                     0.32   \n",
       "10        LPSLPGHTA               9                                     0.00   \n",
       "\n",
       "    Fraction of >80% matching sequences (%)  \\\n",
       "15                                    81.59   \n",
       "5                                     83.17   \n",
       "7                                     83.17   \n",
       "27                                    50.16   \n",
       "22                                    64.13   \n",
       "13                                    82.86   \n",
       "14                                    33.02   \n",
       "3                                     37.46   \n",
       "16                                    70.79   \n",
       "21                                    53.33   \n",
       "26                                    20.00   \n",
       "29                                    15.24   \n",
       "33                                    15.24   \n",
       "34                                    14.92   \n",
       "35                                    14.92   \n",
       "36                                    14.92   \n",
       "17                                    30.16   \n",
       "0                                     33.65   \n",
       "8                                     63.17   \n",
       "30                                    14.29   \n",
       "20                                    24.44   \n",
       "11                                    37.14   \n",
       "19                                    24.44   \n",
       "18                                    17.46   \n",
       "32                                    15.24   \n",
       "2                                     44.76   \n",
       "24                                    11.75   \n",
       "12                                    63.17   \n",
       "6                                     34.60   \n",
       "31                                    14.92   \n",
       "1                                     38.41   \n",
       "4                                     20.95   \n",
       "28                                    59.37   \n",
       "23                                    33.02   \n",
       "25                                     2.22   \n",
       "9                                     56.83   \n",
       "10                                    44.44   \n",
       "\n",
       "    Fraction of <50% matching sequences (%)  \\\n",
       "15                                    15.24   \n",
       "5                                     15.24   \n",
       "7                                     15.24   \n",
       "27                                    15.24   \n",
       "22                                    15.24   \n",
       "13                                     0.63   \n",
       "14                                    15.24   \n",
       "3                                     16.19   \n",
       "16                                    15.24   \n",
       "21                                    15.87   \n",
       "26                                    22.54   \n",
       "29                                    84.76   \n",
       "33                                    84.76   \n",
       "34                                    85.08   \n",
       "35                                    85.08   \n",
       "36                                    85.08   \n",
       "17                                    40.32   \n",
       "0                                     15.24   \n",
       "8                                     15.24   \n",
       "30                                    84.13   \n",
       "20                                    24.76   \n",
       "11                                    21.59   \n",
       "19                                    26.67   \n",
       "18                                    17.46   \n",
       "32                                    84.44   \n",
       "2                                     15.56   \n",
       "24                                    17.46   \n",
       "12                                    15.24   \n",
       "6                                     15.24   \n",
       "31                                    84.76   \n",
       "1                                     15.24   \n",
       "4                                     25.40   \n",
       "28                                    16.19   \n",
       "23                                    16.19   \n",
       "25                                    17.46   \n",
       "9                                     18.10   \n",
       "10                                    15.56   \n",
       "\n",
       "    Fraction of 100% matching in IAV seq (%) # Subtypes 100% conserved  \\\n",
       "15                                     84.62                      8/10   \n",
       "5                                      61.54                      9/10   \n",
       "7                                      50.77                      9/10   \n",
       "27                                     45.77                      8/10   \n",
       "22                                     35.00                      8/10   \n",
       "13                                     34.23                      4/10   \n",
       "14                                     27.69                      2/10   \n",
       "3                                      25.00                      3/10   \n",
       "16                                     20.38                      4/10   \n",
       "21                                     19.62                      2/10   \n",
       "26                                     17.69                      2/10   \n",
       "29                                      0.00                      0/10   \n",
       "33                                      0.00                      0/10   \n",
       "34                                      0.00                      0/10   \n",
       "35                                      0.00                      0/10   \n",
       "36                                      0.00                      0/10   \n",
       "17                                     15.38                      2/10   \n",
       "0                                      15.00                      5/10   \n",
       "8                                      15.00                      5/10   \n",
       "30                                      0.00                      0/10   \n",
       "20                                     13.46                      2/10   \n",
       "11                                     14.23                      2/10   \n",
       "19                                     13.46                      2/10   \n",
       "18                                     13.46                      3/10   \n",
       "32                                      0.00                      0/10   \n",
       "2                                       6.54                      2/10   \n",
       "24                                      3.08                      2/10   \n",
       "12                                      3.08                      3/10   \n",
       "6                                       3.08                      1/10   \n",
       "31                                      0.00                      0/10   \n",
       "1                                       1.54                      2/10   \n",
       "4                                       1.54                      1/10   \n",
       "28                                      1.15                      1/10   \n",
       "23                                      1.15                      1/10   \n",
       "25                                      0.77                      1/10   \n",
       "9                                       0.38                      1/10   \n",
       "10                                      0.00                      0/10   \n",
       "\n",
       "   # subs 80 conserved  Fraction of 100% matching in IBV seq (%)  \n",
       "15               10/10                                      0.00  \n",
       "5                10/10                                      0.00  \n",
       "7                 9/10                                      0.00  \n",
       "27                8/10                                      0.00  \n",
       "22               10/10                                      0.00  \n",
       "13                9/10                                      0.00  \n",
       "14                4/10                                      0.00  \n",
       "3                 4/10                                      0.00  \n",
       "16               10/10                                      0.00  \n",
       "21                9/10                                      0.00  \n",
       "26                3/10                                      0.00  \n",
       "29                0/10                                     97.37  \n",
       "33                0/10                                     94.74  \n",
       "34                0/10                                     97.37  \n",
       "35                0/10                                     97.37  \n",
       "36                0/10                                     97.37  \n",
       "17                4/10                                      0.00  \n",
       "0                 7/10                                      0.00  \n",
       "8                 9/10                                      0.00  \n",
       "30                0/10                                     81.58  \n",
       "20                6/10                                      0.00  \n",
       "11                4/10                                      0.00  \n",
       "19                6/10                                      0.00  \n",
       "18                3/10                                      0.00  \n",
       "32                0/10                                     73.68  \n",
       "2                 7/10                                      0.00  \n",
       "24                2/10                                      0.00  \n",
       "12                8/10                                      0.00  \n",
       "6                 8/10                                      0.00  \n",
       "31                0/10                                     15.79  \n",
       "1                 8/10                                      0.00  \n",
       "4                 5/10                                      0.00  \n",
       "28                8/10                                      0.00  \n",
       "23                3/10                                      0.00  \n",
       "25                3/10                                      0.00  \n",
       "9                 8/10                                      0.00  \n",
       "10                9/10                                      0.00  "
      ]
     },
     "execution_count": 368,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h_conservancy_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 370,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Epitope sequence</th>\n",
       "      <th>Epitope length</th>\n",
       "      <th>Fraction of 100% matching sequences (%)</th>\n",
       "      <th>Fraction of &gt;80% matching sequences (%)</th>\n",
       "      <th>Fraction of &lt;50% matching sequences (%)</th>\n",
       "      <th># Subtypes 100% conserved</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>DRLRRDQKS</td>\n",
       "      <td>9</td>\n",
       "      <td>87.00</td>\n",
       "      <td>99.23</td>\n",
       "      <td>0.00</td>\n",
       "      <td>42/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GEISPLPSL</td>\n",
       "      <td>9</td>\n",
       "      <td>75.39</td>\n",
       "      <td>91.95</td>\n",
       "      <td>0.00</td>\n",
       "      <td>40/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>MSRDWFMLM</td>\n",
       "      <td>9</td>\n",
       "      <td>59.44</td>\n",
       "      <td>82.20</td>\n",
       "      <td>0.00</td>\n",
       "      <td>40/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>GVLIGGLEW</td>\n",
       "      <td>9</td>\n",
       "      <td>47.83</td>\n",
       "      <td>70.90</td>\n",
       "      <td>0.00</td>\n",
       "      <td>34/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>IILKANFSV</td>\n",
       "      <td>9</td>\n",
       "      <td>46.44</td>\n",
       "      <td>96.28</td>\n",
       "      <td>0.15</td>\n",
       "      <td>31/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>IASVPTSRY</td>\n",
       "      <td>9</td>\n",
       "      <td>23.84</td>\n",
       "      <td>56.97</td>\n",
       "      <td>4.33</td>\n",
       "      <td>13/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>VPASRYLTDM</td>\n",
       "      <td>10</td>\n",
       "      <td>17.34</td>\n",
       "      <td>80.96</td>\n",
       "      <td>0.62</td>\n",
       "      <td>26/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>MTIASVPTSRY</td>\n",
       "      <td>11</td>\n",
       "      <td>17.03</td>\n",
       "      <td>66.25</td>\n",
       "      <td>8.67</td>\n",
       "      <td>12/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>MTIASVPTSR</td>\n",
       "      <td>10</td>\n",
       "      <td>17.03</td>\n",
       "      <td>66.41</td>\n",
       "      <td>13.62</td>\n",
       "      <td>12/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>MPRQKIIGPL</td>\n",
       "      <td>10</td>\n",
       "      <td>10.22</td>\n",
       "      <td>26.78</td>\n",
       "      <td>10.22</td>\n",
       "      <td>3/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AIMEKNIIL</td>\n",
       "      <td>9</td>\n",
       "      <td>10.06</td>\n",
       "      <td>29.57</td>\n",
       "      <td>2.01</td>\n",
       "      <td>3/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AIMDKNIIL</td>\n",
       "      <td>9</td>\n",
       "      <td>8.36</td>\n",
       "      <td>47.21</td>\n",
       "      <td>0.31</td>\n",
       "      <td>22/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>IMDKNIILKA</td>\n",
       "      <td>10</td>\n",
       "      <td>8.20</td>\n",
       "      <td>79.57</td>\n",
       "      <td>0.31</td>\n",
       "      <td>22/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>RLRRDQRSL</td>\n",
       "      <td>9</td>\n",
       "      <td>3.25</td>\n",
       "      <td>90.71</td>\n",
       "      <td>0.00</td>\n",
       "      <td>7/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>AIVDKNITL</td>\n",
       "      <td>9</td>\n",
       "      <td>1.70</td>\n",
       "      <td>18.11</td>\n",
       "      <td>2.94</td>\n",
       "      <td>4/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>IMLKANFSV</td>\n",
       "      <td>9</td>\n",
       "      <td>1.70</td>\n",
       "      <td>92.72</td>\n",
       "      <td>0.15</td>\n",
       "      <td>7/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>AVMEKNIVL</td>\n",
       "      <td>9</td>\n",
       "      <td>1.39</td>\n",
       "      <td>4.49</td>\n",
       "      <td>8.36</td>\n",
       "      <td>3/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MASTPASRY</td>\n",
       "      <td>9</td>\n",
       "      <td>0.62</td>\n",
       "      <td>2.48</td>\n",
       "      <td>6.04</td>\n",
       "      <td>1/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AIMDKNIML</td>\n",
       "      <td>9</td>\n",
       "      <td>0.46</td>\n",
       "      <td>21.83</td>\n",
       "      <td>0.31</td>\n",
       "      <td>3/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>KRKMARTAR</td>\n",
       "      <td>9</td>\n",
       "      <td>0.46</td>\n",
       "      <td>0.93</td>\n",
       "      <td>48.30</td>\n",
       "      <td>1/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>QELSDAPFL</td>\n",
       "      <td>9</td>\n",
       "      <td>0.46</td>\n",
       "      <td>31.11</td>\n",
       "      <td>5.11</td>\n",
       "      <td>2/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ETIVLLRAF</td>\n",
       "      <td>9</td>\n",
       "      <td>0.46</td>\n",
       "      <td>7.59</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>IPKQKVAGPL</td>\n",
       "      <td>10</td>\n",
       "      <td>0.31</td>\n",
       "      <td>45.51</td>\n",
       "      <td>13.31</td>\n",
       "      <td>2/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>FLWHVRKRV</td>\n",
       "      <td>9</td>\n",
       "      <td>0.31</td>\n",
       "      <td>54.64</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>AIMEKNIVL</td>\n",
       "      <td>9</td>\n",
       "      <td>0.31</td>\n",
       "      <td>12.69</td>\n",
       "      <td>2.94</td>\n",
       "      <td>2/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>AIMEKNIVLK</td>\n",
       "      <td>10</td>\n",
       "      <td>0.31</td>\n",
       "      <td>50.62</td>\n",
       "      <td>2.94</td>\n",
       "      <td>2/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ALKMTMASV</td>\n",
       "      <td>9</td>\n",
       "      <td>0.31</td>\n",
       "      <td>19.97</td>\n",
       "      <td>14.40</td>\n",
       "      <td>2/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AIMEKNIML</td>\n",
       "      <td>9</td>\n",
       "      <td>0.15</td>\n",
       "      <td>12.07</td>\n",
       "      <td>2.94</td>\n",
       "      <td>1/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>LPSLPGHTA</td>\n",
       "      <td>9</td>\n",
       "      <td>0.15</td>\n",
       "      <td>77.40</td>\n",
       "      <td>0.46</td>\n",
       "      <td>1/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ALDYPGQDRL</td>\n",
       "      <td>10</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>AYDQSGRLV</td>\n",
       "      <td>9</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>99.85</td>\n",
       "      <td>0/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>FMDPSAGIEGF</td>\n",
       "      <td>11</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>ILNSLFERL</td>\n",
       "      <td>9</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>99.38</td>\n",
       "      <td>0/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>NVLSLRVLV</td>\n",
       "      <td>9</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>99.85</td>\n",
       "      <td>0/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>SQFGQEHRL</td>\n",
       "      <td>9</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>SQFGQEHRL</td>\n",
       "      <td>9</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0/43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>VLSQFGQEHRL</td>\n",
       "      <td>11</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0/43</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Epitope sequence  Epitope length  Fraction of 100% matching sequences (%)  \\\n",
       "5         DRLRRDQKS               9                                    87.00   \n",
       "7         GEISPLPSL               9                                    75.39   \n",
       "27        MSRDWFMLM               9                                    59.44   \n",
       "15        GVLIGGLEW               9                                    47.83   \n",
       "22        IILKANFSV               9                                    46.44   \n",
       "26        IASVPTSRY               9                                    23.84   \n",
       "12       VPASRYLTDM              10                                    17.34   \n",
       "20      MTIASVPTSRY              11                                    17.03   \n",
       "19       MTIASVPTSR              10                                    17.03   \n",
       "18       MPRQKIIGPL              10                                    10.22   \n",
       "2         AIMEKNIIL               9                                    10.06   \n",
       "0         AIMDKNIIL               9                                     8.36   \n",
       "8        IMDKNIILKA              10                                     8.20   \n",
       "13        RLRRDQRSL               9                                     3.25   \n",
       "25        AIVDKNITL               9                                     1.70   \n",
       "16        IMLKANFSV               9                                     1.70   \n",
       "24        AVMEKNIVL               9                                     1.39   \n",
       "11        MASTPASRY               9                                     0.62   \n",
       "1         AIMDKNIML               9                                     0.46   \n",
       "17        KRKMARTAR               9                                     0.46   \n",
       "21        QELSDAPFL               9                                     0.46   \n",
       "14        ETIVLLRAF               9                                     0.46   \n",
       "9        IPKQKVAGPL              10                                     0.31   \n",
       "6         FLWHVRKRV               9                                     0.31   \n",
       "23        AIMEKNIVL               9                                     0.31   \n",
       "28       AIMEKNIVLK              10                                     0.31   \n",
       "4         ALKMTMASV               9                                     0.31   \n",
       "3         AIMEKNIML               9                                     0.15   \n",
       "10        LPSLPGHTA               9                                     0.15   \n",
       "29       ALDYPGQDRL              10                                     0.00   \n",
       "30        AYDQSGRLV               9                                     0.00   \n",
       "31      FMDPSAGIEGF              11                                     0.00   \n",
       "32        ILNSLFERL               9                                     0.00   \n",
       "33        NVLSLRVLV               9                                     0.00   \n",
       "34        SQFGQEHRL               9                                     0.00   \n",
       "35        SQFGQEHRL               9                                     0.00   \n",
       "36      VLSQFGQEHRL              11                                     0.00   \n",
       "\n",
       "    Fraction of >80% matching sequences (%)  \\\n",
       "5                                     99.23   \n",
       "7                                     91.95   \n",
       "27                                    82.20   \n",
       "15                                    70.90   \n",
       "22                                    96.28   \n",
       "26                                    56.97   \n",
       "12                                    80.96   \n",
       "20                                    66.25   \n",
       "19                                    66.41   \n",
       "18                                    26.78   \n",
       "2                                     29.57   \n",
       "0                                     47.21   \n",
       "8                                     79.57   \n",
       "13                                    90.71   \n",
       "25                                    18.11   \n",
       "16                                    92.72   \n",
       "24                                     4.49   \n",
       "11                                     2.48   \n",
       "1                                     21.83   \n",
       "17                                     0.93   \n",
       "21                                    31.11   \n",
       "14                                     7.59   \n",
       "9                                     45.51   \n",
       "6                                     54.64   \n",
       "23                                    12.69   \n",
       "28                                    50.62   \n",
       "4                                     19.97   \n",
       "3                                     12.07   \n",
       "10                                    77.40   \n",
       "29                                     0.00   \n",
       "30                                     0.00   \n",
       "31                                     0.00   \n",
       "32                                     0.00   \n",
       "33                                     0.00   \n",
       "34                                     0.00   \n",
       "35                                     0.00   \n",
       "36                                     0.00   \n",
       "\n",
       "    Fraction of <50% matching sequences (%) # Subtypes 100% conserved  \n",
       "5                                      0.00                     42/43  \n",
       "7                                      0.00                     40/43  \n",
       "27                                     0.00                     40/43  \n",
       "15                                     0.00                     34/43  \n",
       "22                                     0.15                     31/43  \n",
       "26                                     4.33                     13/43  \n",
       "12                                     0.62                     26/43  \n",
       "20                                     8.67                     12/43  \n",
       "19                                    13.62                     12/43  \n",
       "18                                    10.22                      3/43  \n",
       "2                                      2.01                      3/43  \n",
       "0                                      0.31                     22/43  \n",
       "8                                      0.31                     22/43  \n",
       "13                                     0.00                      7/43  \n",
       "25                                     2.94                      4/43  \n",
       "16                                     0.15                      7/43  \n",
       "24                                     8.36                      3/43  \n",
       "11                                     6.04                      1/43  \n",
       "1                                      0.31                      3/43  \n",
       "17                                    48.30                      1/43  \n",
       "21                                     5.11                      2/43  \n",
       "14                                     0.00                      2/43  \n",
       "9                                     13.31                      2/43  \n",
       "6                                      0.00                      2/43  \n",
       "23                                     2.94                      2/43  \n",
       "28                                     2.94                      2/43  \n",
       "4                                     14.40                      2/43  \n",
       "3                                      2.94                      1/43  \n",
       "10                                     0.46                      1/43  \n",
       "29                                   100.00                      0/43  \n",
       "30                                    99.85                      0/43  \n",
       "31                                   100.00                      0/43  \n",
       "32                                    99.38                      0/43  \n",
       "33                                    99.85                      0/43  \n",
       "34                                   100.00                      0/43  \n",
       "35                                   100.00                      0/43  \n",
       "36                                   100.00                      0/43  "
      ]
     },
     "execution_count": 370,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "avs_conservancy_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 371,
   "metadata": {},
   "outputs": [],
   "source": [
    "h_conservancy_table.to_csv(\"../epitope_conservancy/{}_AB_clust{}_conservancy.csv\".format(prot, h_clust_str, index=False))\n",
    "avs_conservancy_table.to_csv(\"../epitope_conservancy/{}_AS_clust{}_conservancy.csv\".format(prot, avs_clust_str, index=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Candidate epitopes detection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read MSA files\n",
    "human_msa = \"../sequences/msa/msa_{}_AB_clust{}.fa\".format(prot, h_clust_str)\n",
    "avswine_msa =  \"../sequences/msa/msa_{}_AS_clust{}.fa\".format(prot, avs_clust_str)\n",
    "\n",
    "msa_hseq = AlignIO.read(human_msa, 'fasta')\n",
    "msa_avseq = AlignIO.read(avswine_msa, 'fasta')\n",
    "\n",
    "# Convert strings of sequences into list of amino acids. (1-Dimensional array -> 2-Dimensional array)\n",
    "msa_hseq, msa_avseq = np.array(msa_hseq), np.array(msa_avseq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Looking for 9-mer candidates . . .\n",
      "Looking for 10-mer candidates . . .\n",
      "Looking for 11-mer candidates . . .\n"
     ]
    }
   ],
   "source": [
    "# Look into MSA - look for candidates of epitope sequences\n",
    "k_mer = [k for k in range(9, 12)]\n",
    "\n",
    "h_stds, h_scores, h_seqs, h_poss = [], [], [], []\n",
    "avs_stds, avs_scores, avs_seqs, avs_poss = [], [], [], []\n",
    "\n",
    "for k in k_mer:\n",
    "    print(\"Looking for {}-mer candidates . . .\".format(k))\n",
    "    h_std, h_score, h_seq, h_pos = find_experimental_epitopes(k, msa_hseq, threshold=0.8)\n",
    "    h_stds.extend(h_std)\n",
    "    h_scores.extend(h_score)\n",
    "    h_seqs.extend(h_seq)\n",
    "    h_poss.extend(h_pos)\n",
    "    \n",
    "    avs_std, avs_score, avs_seq, avs_pos = find_experimental_epitopes(k, msa_avseq, threshold=0.8)\n",
    "    avs_stds.extend(avs_std)\n",
    "    avs_scores.extend(avs_score)\n",
    "    avs_seqs.extend(avs_seq)\n",
    "    avs_poss.extend(avs_pos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing human sequences . . .\n",
      "Processing avian-swine sequences . . .\n"
     ]
    }
   ],
   "source": [
    "# Make dataframe of results!\n",
    "print(\"Processing human sequences . . .\")\n",
    "h_exp_results = exp_epitopes_dataframe(h_stds, h_scores, h_seqs)\n",
    "print(\"Processing avian-swine sequences . . .\")\n",
    "avs_exp_results = exp_epitopes_dataframe(avs_stds, avs_scores, avs_seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "avs_exp_nonoverlap = non_overlapping_epitopes_simple(avs_seqs, epitopes)\n",
    "h_exp_nonoverlap = non_overlapping_epitopes_simple(h_seqs, epitopes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>candidate sequence</th>\n",
       "      <th>length</th>\n",
       "      <th>substitution score</th>\n",
       "      <th>best case</th>\n",
       "      <th>score (%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [candidate sequence, length, substitution score, best case, score (%)]\n",
       "Index: []"
      ]
     },
     "execution_count": 207,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h_exp_results_nonoverlap = h_exp_results.loc[h_exp_results['candidate sequence'].isin(h_exp_nonoverlap)]\n",
    "h_exp_results_nonoverlap.loc[h_exp_results_nonoverlap['score (%)'] > 99]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>candidate sequence</th>\n",
       "      <th>length</th>\n",
       "      <th>substitution score</th>\n",
       "      <th>best case</th>\n",
       "      <th>score (%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>KFEEIRWLI</td>\n",
       "      <td>9</td>\n",
       "      <td>24338.0</td>\n",
       "      <td>24598.0</td>\n",
       "      <td>98.94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>QKFEEIRWLI</td>\n",
       "      <td>10</td>\n",
       "      <td>26797.0</td>\n",
       "      <td>27108.0</td>\n",
       "      <td>98.85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>QKFEEIRWL</td>\n",
       "      <td>9</td>\n",
       "      <td>24807.0</td>\n",
       "      <td>25100.0</td>\n",
       "      <td>98.83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>QALQLLLEVE</td>\n",
       "      <td>10</td>\n",
       "      <td>21830.0</td>\n",
       "      <td>22088.0</td>\n",
       "      <td>98.83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>QALQLLLEV</td>\n",
       "      <td>9</td>\n",
       "      <td>19341.0</td>\n",
       "      <td>19578.0</td>\n",
       "      <td>98.79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>DLHSLQSRNGK</td>\n",
       "      <td>11</td>\n",
       "      <td>23842.0</td>\n",
       "      <td>28614.0</td>\n",
       "      <td>83.32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>SLQSRNGKW</td>\n",
       "      <td>9</td>\n",
       "      <td>20645.0</td>\n",
       "      <td>25100.0</td>\n",
       "      <td>82.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>LHSLQSRNG</td>\n",
       "      <td>9</td>\n",
       "      <td>18903.0</td>\n",
       "      <td>23092.0</td>\n",
       "      <td>81.86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>LHSLQSRNGK</td>\n",
       "      <td>10</td>\n",
       "      <td>20885.0</td>\n",
       "      <td>25602.0</td>\n",
       "      <td>81.58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>HSLQSRNGK</td>\n",
       "      <td>9</td>\n",
       "      <td>19123.0</td>\n",
       "      <td>23594.0</td>\n",
       "      <td>81.05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>228 rows  5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    candidate sequence  length  substitution score  best case  score (%)\n",
       "38           KFEEIRWLI       9             24338.0    24598.0      98.94\n",
       "116         QKFEEIRWLI      10             26797.0    27108.0      98.85\n",
       "37           QKFEEIRWL       9             24807.0    25100.0      98.83\n",
       "146         QALQLLLEVE      10             21830.0    22088.0      98.83\n",
       "67           QALQLLLEV       9             19341.0    19578.0      98.79\n",
       "..                 ...     ...                 ...        ...        ...\n",
       "176        DLHSLQSRNGK      11             23842.0    28614.0      83.32\n",
       "23           SLQSRNGKW       9             20645.0    25100.0      82.25\n",
       "21           LHSLQSRNG       9             18903.0    23092.0      81.86\n",
       "100         LHSLQSRNGK      10             20885.0    25602.0      81.58\n",
       "22           HSLQSRNGK       9             19123.0    23594.0      81.05\n",
       "\n",
       "[228 rows x 5 columns]"
      ]
     },
     "execution_count": 208,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "avs_exp_results_nonoverlap = avs_exp_results.loc[avs_exp_results['candidate sequence'].isin(avs_exp_nonoverlap)]\n",
    "avs_exp_results_nonoverlap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export the results \n",
    "h_exp_results_nonoverlap.to_csv(\"../sequences/nonoverlap_candidates/{}_AB_clust{}_candidate_epitopes.csv\".format(prot, h_clust_str, index=False))\n",
    "avs_exp_results_nonoverlap.to_csv(\"../sequences/nonoverlap_candidates/{}_AS_clust{}_candidate_epitopes.csv\".format(prot, avs_clust_str, index=False))\n",
    "\n",
    "h_conservancy_table.to_csv(\"../sequences/epitope_conservancy/{}_AB_clust{}_conservancy.csv\".format(prot, h_clust_str, index=False))\n",
    "avs_conservancy_table.to_csv(\"../sequences/epitope_conservancy/{}_AS_clust{}_conservancy.csv\".format(prot, avs_clust_str, index=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "1bd06f5ea03ec233ea613593a8abdc9f0315a389ccaf2775c94be72fffa4b7be"
  },
  "kernelspec": {
   "display_name": "Python 3.8.12 64-bit ('base': conda)",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
